ot/torch/lp.py

```diff
@@ -79,3 +78,95 @@ def ot_solve(a, b, M, num_iter_max=100000, log=False):
         G = emd(a2, b2, M2, log=False, numItermax=num_iter_max)
 
         return torch.from_numpy(G).type_as(M)
+
+
+def ot_loss_1d(u_values, v_values, u_weights=None, v_weights=None, p=1, require_sort=True):
+    r"""
+    Computes the 1 dimensional OT loss [2] between two (batched) empirical distributions
+    ..math:
+        ot_{loss} &= \int_0^1 |cdf_u^{-1}(q)  cdf_v^{-1}(q)|^p dq
+
+    It is formally the p-Wasserstein distance raised to the power p.
+    We do so in a vectorized way by first building the individual quantile functions then integrating them.
+    This has a theoretically higher complexity than the core OT implementation but behaves better with PyTorch
+
+    Parameters
+    ----------
+    u_values: torch.Tensor (n, ...)
+        locations of the first empirical distribution
+    v_values: torch.Tensor (m, ...)
+        locations of the second empirical distribution
+    u_weights: torch.Tensor (n, ...), optional
+        weights of the first empirical distribution, if None then uniform weights are used
+    v_weights: torch.Tensor (m, ...), optional
+        weights of the second empirical distribution, if None then uniform weights are used
+    p: int, optional
+        order of the ground metric used, should be at least 1 (see [2, Chap. 2], default is 1
+    require_sort: bool, optional
+        Are the locations sorted along the last dimension already
+
+    Returns
+    -------
+    cost: torch.Tensor (...,)
+        the batched EMD
+
+    Examples
+    --------
+    Simple example:
+    >>> import ot
+    >>> import torch
+    >>> np.random.seed(0)
+    >>> n_source = 7
+    >>> n_target = 100
+    >>> a = torch.tensor(ot.utils.unif(n_source), requires_grad=True)
+    >>> b = torch.tensor(ot.utils.unif(n_target))
+    >>> X_source = torch.tensor(np.random.randn(n_source,), requires_grad=True)
+    >>> Y_target = torch.tensor(np.random.randn(n_target,))
+    >>> loss = ot.torch.lp.ot_loss_1d(X_source, Y_target, a, b)
+    >>> torch.autograd.grad(loss, X_source)[0]
+    tensor([0.1429, 0.1429, 0.1429, 0.1229, 0.1429, 0.1429, 0.1429],
+           dtype=torch.float64)
+
+    References
+    ----------
+    .. [2] Cuturi, M. (2013). [Sinkhorn distances: Lightspeed computation of optimal transport](https://arxiv.org/pdf/1306.0895.pdf). In Advances in Neural Information Processing Systems (pp. 2292-2300).
+
+    """
+    assert p >= 1, "The OT loss is only valid for p>=1, {p} was given".format(p=p)
```

Comment:
should we also assert that all tensors are on  the same device, and share the same dtype ? It is an open question, I do  not know what the common practices are for handling this type of error

Reply:
I would tend to let Pytorch handle that kind of things... They probably would do it much better than I would